language/move-vm/runtime/src/loader.rs

```diff
@@ -0,0 +1,1378 @@
+    fn get(&self, key: &K) -> Option<&Arc<V>> {
+        self.id_map
+            .get(&key)
+            .and_then(|idx| self.binaries.get(*idx))
+    }
+}
+
+// A script cache is a map from the hash value of a script and the `Script` itself.
+// Script are added in the cache once verified and so getting a script out the cache
+// does not require further verification (except for parameters and type parameters)
+struct ScriptCache {
+    scripts: BinaryCache<HashValue, Script>,
+}
+
+impl ScriptCache {
+    fn new() -> Self {
+        Self {
+            scripts: BinaryCache::new(),
+        }
+    }
+
+    fn get(&self, hash: &HashValue) -> Option<Arc<Function>> {
+        self.scripts.get(hash).map(|script| script.entry_point())
+    }
+
+    fn insert(&mut self, hash: HashValue, script: Script) -> VMResult<Arc<Function>> {
+        match self.get(&hash) {
+            Some(script) => Ok(script),
+            None => Ok(self.scripts.insert(hash, script).entry_point()),
+        }
+    }
+}
+
+// A ModuleCache is the core structure in the Loader.
+// It holds all Modules, Types and Functions loaded.
+// Types and Functions are pushed globally to the ModuleCache.
+// A ModuleCache is accessed under lock.
+struct ModuleCache {
+    modules: BinaryCache<ModuleId, Module>,
+    structs: Vec<Arc<StructType>>,
+    functions: Vec<Arc<Function>>,
+}
+
+impl ModuleCache {
+    fn new() -> Self {
+        Self {
+            modules: BinaryCache::new(),
+            structs: vec![],
+            functions: vec![],
+        }
+    }
+
+    fn get(&self, id: &ModuleId) -> Option<Arc<Module>> {
+        self.modules.get(id).map(|module| Arc::clone(module))
+    }
+
+    fn insert(&mut self, id: ModuleId, ver_module: VerifiedModule) -> VMResult<Arc<Module>> {
+        self.publish_module(&ver_module)?;
+        let module = Module::new(ver_module, self)?;
+        match self.get(&id) {
+            Some(module) => Ok(module),
+            None => Ok(Arc::clone(self.modules.insert(id, module))),
+        }
+    }
+
+    fn function_at(&self, idx: usize) -> Arc<Function> {
+        self.functions[idx].clone()
+    }
+
+    fn struct_at(&self, idx: usize) -> Arc<StructType> {
+        Arc::clone(&self.structs[idx])
+    }
+
+    fn publish_module(&mut self, module: &VerifiedModule) -> VMResult<()> {
+        let starting_idx = self.structs.len();
+        for (idx, struct_def) in module.struct_defs().iter().enumerate() {
+            let st = self.load_type(module, struct_def, StructDefinitionIndex(idx as u16))?;
+            self.structs.push(Arc::new(st));
+        }
+        self.load_fields(module, starting_idx)?;
+
+        for func in module.function_defs() {
+            let function = self.load_function(module, func)?;
+            self.functions.push(Arc::new(function));
+        }
+        Ok(())
+    }
+
+    fn load_type(
+        &self,
+        module: &VerifiedModule,
+        struct_def: &StructDefinition,
+        idx: StructDefinitionIndex,
```

Comment:
You don't need to pass in the `&StructDefinition` since you already have the index?

Reply:
I could refactor the code a bit but I will not right now. The outer loop loads the index and the definition and they are passed in. We could loop through the indexes only and then load the definition here but I am not sure it makes any difference or even for better code.
Fair point though... just cannot convince myself any other solution is better for now.
Something to review